Core/include/Acts/Detector/LayerStructureBuilder.hpp

```diff
@@ -0,0 +1,105 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Detector/IInternalStructureBuilder.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+#include <functional>
+#include <optional>
+
+namespace Acts {
+namespace Experimental {
+
+/// @brief This is a builder of layer structures to be contained
+/// within a DetectorVolume.
+///
+/// It uses the IndexedSurfaceGrid to bin the internal surfaces,
+/// and allows for additional support surfaces that are added to the
+/// structure and indexing mechanism. Those support structures can
+/// also be approximated by planar surfaces, in order to facilitate
+/// vectorization of surface intersection calls.
+///
+/// No sub volumes are added to this structure builders, hence,
+/// the DetectorVolumeFinder navigation delegate uses the "NoopFinder"
+/// breakpoint to indicate the bottom of the volume hierarchy.
+///
+class LayerStructureBuilder : public IInternalStructureBuilder {
+ public:
+  /// @brief Support parameter defintions
+  struct Support {
+    /// Define whether you want to build support structures
+    std::array<ActsScalar, 5u> values = {};
+    /// The surface type to be built
+    Surface::SurfaceType type = Surface::SurfaceType::Other;
+    /// Define in which values the support should be constrained
+    std::vector<BinningValue> constraints = s_binningValues;
+    /// Potential splits into planar approximations
+    unsigned int splits = 1u;
+    /// The (optional) layer transform
+    std::optional<Transform3> transform = std::nullopt;
+  };
+
+  /// @brief The surface binning definition
+  struct Binning {
+    /// Define the binning
+    BinningData data;
+    /// An expansion for the filling
+    size_t expansion = 0u;
+  };
+
+  /// @brief Configuration struct for the LayerStructureBuilder
+  ///
+  /// It contain:
+  /// - a source of the surfaces to be built
+  /// - a definition of surface binning on this layer
+  /// - a definition of supports to be built
+  struct Config {
+    /// Connection point for a function to provide surfaces
+    std::function<std::vector<std::shared_ptr<Surface>>()> surfaces;
```

Comment:
should we typedef the function type?

Reply:
That's a vector os surfaces, I think i'ts fine ...